README
DB-Color

Syntax highlighting for the Perl debugger.

Want this?
http://blogs.perl.org/users/ovid/2010/11/syntax-highlighting-for-the-perl-debugger.html

Then add this to your $HOME/.perldb

 use DB::Color;

Then use your debugger like normal:

 perl -d some_file.pl

If you don't want a $HOME/.perldb file:

 perl -MDB::Color -d some_file.pl

See perldoc DB::Color for more information.

This code is a proof of concept and is buggy. The syntax highlighting is
buggy. Fix it :)

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc DB::Color

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=DB-Color

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/DB-Color

    CPAN Ratings
        http://cpanratings.perl.org/d/DB-Color

    Search CPAN
        http://search.cpan.org/dist/DB-Color/


LICENSE AND COPYRIGHT

Copyright (C) 2011 Curtis "Ovid" Poe

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for DB-Color

0.20   2015/11/21
        - Add optional sentinel support to avoid loading DB::Color unless
          requested.
        - Add an example in examples/colorize of a program which can
          continuously colorize your files on the fly.
        - Add a "WORKFLOW" section to the docs to explain more efficient usage
          of this module.
        - Remove Module::Build. RIP :/

0.10   2012/01/08
        - Fix "Argument isn't numeric" warnings.
        - Assert minimum version of File::Path to fix "make_path is not
          exported" build failures.

0.09   2012/09/05
        - Don't use BRIGHT_BLUE if they have Term::ANSIColor < 3 (Cosimo
          Streppone)

0.08   2012/04/08
        - Fixed bug where two or more package declarations in a single file
          would break syntax highlighting.
        - Added a perldbsyntax script to pre-generate syntax files.

0.07   2012/04/01
        - Add a config file to let people control some behavior.
        - The afterinit() sub is no longer required in the .perldb file (it's
          done internally).
        - The md5 sum is now generate with the "format number" and classname.
        - Caching can now be disabled by not passing a cache_dir to the
          highlighter.

0.06    2012/03/31
        - Don't highlight if NO_DB_COLOR environment variable is set to a true value.
        - Add optional debugging to log file (internal and undocumented. Use
          at own risk.
        - Naive caching of output files. File deleted after thirty days of
          non-use.
        - Workaround Syntax::Highlight::Kate POD bug (https://rt.cpan.org/Ticket/Display.html?id=76160)
        - Fix broken breakpoints. Thanks to Nick Perez and Liz for the help!
        - Fix off-by-one error (Thanks Liz!)

0.05    2012/01/04
        - Ouch! Add the actual highlighter to the MANIFEST

0.04    2012/01/04
        - Fix Github link in Build.PL

0.03    2012/01/04
        - Add Build.PL. List github repo.
        - Syntax highlighter requires 5.8.0, so put that in the Makefile.PL
        - POD for DB::Color::Highlighter.

0.02    2011/12/30
        - Quick hack to silence uninit warnings in DB::DB
        - Silence warnings from Kate highlighter
        - Tried to make the colors easier to work with

0.01    2011/12/30
        - Debugger syntax highlighting



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use inc::Module::Install;

name 'DB-Color';
all_from 'lib/DB/Color.pm';
author q{Curtis "Ovid" Poe <ovid@cpan.org>};
license 'artistic2';

perl_version 5.006;

tests_recursive('t');

resources(
    homepage   => 'https://github.com/Ovid/DB--Color',
    license    => 'http://www.perlfoundation.org/artistic_license_2_0',
    repository => 'https://github.com/Ovid/DB--Color',
    bugtracker => 'https://github.com/Ovid/DB--Color/issues',
);

configure_requires(
    'Module::Install' => 0,
);

build_requires(
    'Test::More' => '0',
);

requires(
    'Syntax::Highlight::Engine::Kate' => 0.06,
    'Digest::MD5'                     => 0,
    'Term::ANSIColor' => 0,        # 3.0 or better to get bright colors
    'File::Path'      => '2.07',
);

install_as_cpan;
auto_install;
WriteAll;


lib/DB/Color.pm
package DB::Color;

use 5.008;
use strict;
use warnings;
use DB::Color::Highlight;
use DB::Color::Config;

use IO::Handle;
use File::Spec::Functions qw(catfile catdir);
use Scalar::Util 'dualvar';
use File::Find;

=head1 NAME

DB::Color - Colorize your debugger output

=head1 VERSION

Version 0.20

=cut

our $VERSION = '0.20';

=head1 SYNOPSIS

Put the following in your F<$HOME/.perldb> file:

 use DB::Color;

Then use your debugger like normal:

 perl -d some_file.pl

If you don't want a F<$HOME/.perldb> file, you can do this:

 perl -MDB::Color -d some_file.pl

=head1 DISABLING COLOR

If the NO_DB_COLOR environment variable is set to a true value, syntax
highlighting will be disabled.

=head1 WINDOWS

No, sorry. It's a combination of bad Windows support for ANSI escape sequences
and bad debugger design.

=head1 PERFORMANCE

When using the debugger and when you step into something, or continue to a
breakpoint in a new file, the debugger may appear to hang for a moment
(perhaps a long moment if the file is big) while the file is syntax
highlighted and cached. The next time the debugger enters this file, the
highlighting should be instantaneous.

You can speed up the debugger by using the L<perldbsyntax> program which is
included in this distribution. It will pregenerate syntax files for you.

Syntax highlighting the code is very slow. As a result, we cache the output
files in F<$HOME/.perldbcolor>. This is done by calculating the md5 sum of the
file contents. If the file is changed, we get a new sum. This means that
syntax highlighting is very slow at first, but every time you hit the same
file, assuming its unchanged, the cached version is served first.

Note that the cache files are removed after they become 30 (but see config)
days old without being used. If you use the debugger regularly, commonly
debugged files will load very quickly (assuming they haven't changed).

=head1 CONDITIONAL LOADING

If you prefer, you may only want to have I<some> of your projects "colorized".
If so, you can do something like this:

    use DB::Color sentinel => '.colorize';

If an if the C<.colorize> sentinel (or whatever you named it) does not exist,
C<DB::Color> will not be used.

=head1 WORKFLOW

To use C<DB::Color> effectively, I recommend the following:

    $ cpanm DB::Color
    $ echo "use DB::Color sentinel => '.colorize'" >> ~/.perldb
    # cd to project you want to colorize and create the sentinel
    $ touch .colorize
    # colorize the project. This will likely take a long time
    $ PERL5LIB=lib:t/tests perldbsyntax

At that point, you're almost good to go. However, as you're rapidly changing
files, the debugger will still probably be very slow. Instead, create a
watcher to watch your project directories and rehighlight any files which have
been created or modified. An example of a watcher program is the
F<examples/colorize> program included with this distribution.

=head1 CONFIGURATION

You can optionally configure C<DB::Color> by creating a
F<$HOME/.perldbcolorrc> configuration file. It looks like this:

 [core]
 
 # the class that will highlight the code
 highlighter = DB::Color::Highlight
 
 # Any cache file not accessed after this number of days is purged
 cache_max_age = 30
 
 # where to put the cache dir
 cache_dir   = /users/ovid/.perldbcolor
 
The above values are more or less the defaults for this module. They are all
optional.

=head1 ALPHA

This is only a proof of concept. In fact, it's fair to say that this code
sucks. It's not very configurable and has bugs. It's also going to possibly be
a memory hog, as if the debugger wasn't bad enough already.

=cut

my $config = DB::Color::Config->read( default_rcfile() );

my %COLORED;
my $DB_BASE_DIR = $config->{core}{cache_dir} || default_base_dir();

my $DB_LOG = catfile( $DB_BASE_DIR, 'debug.log' );
my $CACHE_MAX_AGE = $config->{core}{cache_max_age} || 30;
my $DEBUG;

# Not documenting this because I don't guarantee stability, but you can play
# with it if you want.
if ( $ENV{DB_COLOR_DEBUG} ) {
    open $DEBUG, '>>', $DB_LOG
      or die "Cannot open $DB_LOG for appending: $!";
    $DEBUG->autoflush(1);
}

my $HIGHLIGHTER_CLASS = $config->{core}{highlighter} || 'DB::Color::Highlight';
eval "use $HIGHLIGHTER_CLASS";
die $@ if $@;

my $HIGHLIGHTER = $HIGHLIGHTER_CLASS->new(
    {
        cache_dir => $DB_BASE_DIR,
        debug_fh  => $DEBUG,
    }
);

sub DB::afterinit {
    no warnings 'once';
    push @DB::typeahead => "{{v"
      unless $DB::already_curly_curly_v++;
}

sub default_rcfile { catfile( $ENV{HOME}, '.perldbcolorrc' ) }
sub default_base_dir { catfile( $ENV{HOME}, '.perldbcolor' ) }

sub import {
    my ( $package, %arg_for ) = @_;
    my $sentinel = $arg_for{sentinel};

    if ( defined $sentinel && !-e $sentinel ) {
        warn "DB::Color not running because '$sentinel' was requested, but not found\n";
        return;
    }
    return if $ENV{NO_DB_COLOR};
    if ( 'MSWin32' eq $^O ) {
        warn <<"END";
DB::Color does not run under Windows because the Windows terminal is too
broken to understand terminal color code.

DB::Color does not use Win32::Console because the debugger is too broken to be
properly extensible.
END
        return;
    }
    my $old_db = \&DB::DB;

    my $new_DB = sub {
        my $lvl = 0;
        while ( my ($pkg) = caller( $lvl++ ) ) {
            return if $pkg eq "DB" or $pkg =~ /^DB::/;
        }
        my ( $package, $filename ) = caller;
        if ($DEBUG) {
            print $DEBUG "In package '$package', filename '$filename'\n";
        }

        # syntax highlight everything and cache it
        my $lines = $COLORED{$filename} ||= do {
            no strict 'refs';
            no warnings 'uninitialized';
            [
                split /(?<=\n)/ =>
                  $HIGHLIGHTER->highlight_text( join "" => @{"::_<$filename"} )
            ];
        };

        {

            # lie to the debugger about what the lines of code are
            no strict 'refs';
            my $line_num = 0;
            foreach ( @{"::_<$filename"} ) {

                # uncomment these to blow your f'in mind
                #if ( not defined ) {
                #    use Devel::Peek;
                #    warn "line number is $line_num";
                #    Dump($_);
                #}
                # The debugger special cases the first value in ::_<$filename.
                # It's "undef" but sometimes contains some data about the
                # program. I don't know entirely what it is, but this solves
                # the "off by one" bug.
                next unless defined;    # thanks Liz! (why does this work?)
                my $line = $lines->[ $line_num++ ];
                next unless defined $line;    # happens when $_ = "\n"

                # XXX Cheap hack to fix
                # Argument "{\n" isn't numeric in addition (+) at DB/Color.pm line 189.
                no warnings 'numeric'; 
                my $numeric_value = 0 + $_;

                # Internally, the debugger uses dualvars for each line of
                # code. If it's numeric value is 0, then the line is not
                # breakable. If we don't include this, no lines in the
                # debugger are breakable.
                $_ = dualvar $numeric_value, $line;
            }
        }
        goto $old_db;
    };

    {
        no warnings 'redefine';
        *DB::DB = $new_DB;
    }

    return;
}

END {
    find(
        sub {

            # delete empty files or files > $CACHE_MAX_AGE days old
            if ( -f $_ && ( -z _ || -M _ > $CACHE_MAX_AGE ) ) {
                unlink($_) or die "Could not unlink '$File::Find::name': $!";
            }
        },
        $DB_BASE_DIR,
    );
    # we're not testing for failure as this is a cheap hack to delete empty
    # directories
    finddepth( sub { rmdir $_ if -d }, $DB_BASE_DIR );
}

1;

=head1 AUTHOR

Curtis "Ovid" Poe, C<< <ovid at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Ovid/DB--Color/issues>.  I will be notified, and then
you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc DB::Color

You can also look for information at:

=over 4

=item * Bug tracker (report bugs here)

L<https://github.com/Ovid/DB--Color/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DB-Color>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DB-Color>

=item * Search CPAN

L<http://search.cpan.org/dist/DB-Color/>

=item * Github

L<http://github.com/Ovid/DB--Color>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to Nick Perez, Liz, and the 2012 Perl Hackathon for helping to overcome
some major hurdles with this module.

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Curtis "Ovid" Poe.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;    # End of DB::Color


lib/DB/Color/Config.pm
package DB::Color::Config;

# If you thought Config::Simple was small...

use strict;
BEGIN {
    require 5.004;
    $DB::Color::Config::VERSION = '0.20';
    $DB::Color::Config::errstr  = '';
}

# Create an empty object
sub new { bless {}, shift }

# Create an object from a file
sub read {
    my $class = ref $_[0] ? ref shift : shift;

    # Check the file
    my $file = shift or return $class->_error( 'You did not specify a file name' );
    return $class->_error( "File '$file' does not exist" )              unless -e $file;
    return $class->_error( "'$file' is a directory, not a file" )       unless -f _;
    return $class->_error( "Insufficient permissions to read '$file'" ) unless -r _;

    # Slurp in the file
    local $/ = undef;
    open( CFG, $file ) or return $class->_error( "Failed to open file '$file': $!" );
    my $contents = <CFG>;
    close( CFG );

    $class->read_string( $contents );
}

# Create an object from a string
sub read_string {
    my $class = ref $_[0] ? ref shift : shift;
    my $self  = bless {}, $class;
    return undef unless defined $_[0];

    # Parse the file
    my $ns      = '_';
    my $counter = 0;
    foreach ( split /(?:\015{1,2}\012|\015|\012)/, shift ) {
        $counter++;

        # Skip comments and empty lines
        next if /^\s*(?:\#|\;|$)/;

        # Remove inline comments
        s/\s\;\s.+$//g;

        # Handle section headers
        if ( /^\s*\[\s*(.+?)\s*\]\s*$/ ) {
            # Create the sub-hash if it doesn't exist.
            # Without this sections without keys will not
            # appear at all in the completed struct.
            $self->{$ns = $1} ||= {};
            next;
        }

        # Handle properties
        if ( /^\s*([^=]+?)\s*=\s*(.*?)\s*$/ ) {
            $self->{$ns}->{$1} = $2;
            next;
        }

        return $self->_error( "Syntax error at line $counter: '$_'" );
    }

    $self;
}

# Save an object to a file
sub write {
    my $self = shift;
    my $file = shift or return $self->_error(
        'No file name provided'
        );

    # Write it to the file
    my $string = $self->write_string;
    return undef unless defined $string;
    open( CFG, '>' . $file ) or return $self->_error(
        "Failed to open file '$file' for writing: $!"
        );
    print CFG $string;
    close CFG;
}

# Save an object to a string
sub write_string {
    my $self = shift;

    my $contents = '';
    foreach my $section ( sort { (($b eq '_') <=> ($a eq '_')) || ($a cmp $b) } keys %$self ) {
        # Check for several known-bad situations with the section
        # 1. Leading whitespace
        # 2. Trailing whitespace
        # 3. Newlines in section name
        return $self->_error(
            "Illegal whitespace in section name '$section'"
        ) if $section =~ /(?:^\s|\n|\s$)/s;
        my $block = $self->{$section};
        $contents .= "\n" if length $contents;
        $contents .= "[$section]\n" unless $section eq '_';
        foreach my $property ( sort keys %$block ) {
            return $self->_error(
                "Illegal newlines in property '$section.$property'"
            ) if $block->{$property} =~ /(?:\012|\015)/s;
            $contents .= "$property=$block->{$property}\n";
        }
    }
    
    $contents;
}

# Error handling
sub errstr { $DB::Color::Config::errstr }
sub _error { $DB::Color::Config::errstr = $_[1]; undef }

1;

__END__

=pod

=head1 NAME

DB::Color::Config - Read/Write .ini style files with as little code as possible

=head1 NOTE

This is an embedded fork of L<Config::Tiny> version 2.14. There is no
functional change.

=head1 SYNOPSIS

    # In your configuration file
    rootproperty=blah

    [section]
    one=twp
    three= four
    Foo =Bar
    empty=

    # In your program
    use DB::Color::Config;

    # Create a config
    my $Config = DB::Color::Config->new;

    # Open the config
    $Config = DB::Color::Config->read( 'file.conf' );

    # Reading properties
    my $rootproperty = $Config->{_}->{rootproperty};
    my $one = $Config->{section}->{one};
    my $Foo = $Config->{section}->{Foo};

    # Changing data
    $Config->{newsection} = { this => 'that' }; # Add a section
    $Config->{section}->{Foo} = 'Not Bar!';     # Change a value
    delete $Config->{_};                        # Delete a value or section

    # Save a config
    $Config->write( 'file.conf' );

=head1 DESCRIPTION

C<DB::Color::Config> is a perl class to read and write .ini style configuration
files with as little code as possible, reducing load time and memory
overhead. Most of the time it is accepted that Perl applications use a lot
of memory and modules. The C<::Tiny> family of modules is specifically
intended to provide an ultralight alternative to the standard modules.

This module is primarily for reading human written files, and anything we
write shouldn't need to have documentation/comments. If you need something
with more power move up to L<Config::Simple>, L<Config::General> or one of
the many other C<Config::> modules. To rephrase, L<DB::Color::Config> does B<not>
preserve your comments, whitespace, or the order of your config file.

=head1 CONFIGURATION FILE SYNTAX

Files are the same format as for windows .ini files. For example:

    [section]
    var1=value1
    var2=value2

If a property is outside of a section at the beginning of a file, it will
be assigned to the C<"root section">, available at C<$Config-E<gt>{_}>.

Lines starting with C<'#'> or C<';'> are considered comments and ignored,
as are blank lines.

When writing back to the config file, all comments, custom whitespace,
and the ordering of your config file elements is discarded. If you need
to keep the human elements of a config when writing back, upgrade to
something better, this module is not for you.

=head1 METHODS

=head2 new

The constructor C<new> creates and returns an empty C<DB::Color::Config> object.

=head2 read $filename

The C<read> constructor reads a config file, and returns a new
C<DB::Color::Config> object containing the properties in the file. 

Returns the object on success, or C<undef> on error.

When C<read> fails, C<DB::Color::Config> sets an error message internally
you can recover via C<DB::Color::Config-E<gt>errstr>. Although in B<some>
cases a failed C<read> will also set the operating system error
variable C<$!>, not all errors do and you should not rely on using
the C<$!> variable.

=head2 read_string $string;

The C<read_string> method takes as argument the contents of a config file
as a string and returns the C<DB::Color::Config> object for it.

=head2 write $filename

The C<write> method generates the file content for the properties, and
writes it to disk to the filename specified.

Returns true on success or C<undef> on error.

=head2 write_string

Generates the file content for the object and returns it as a string.

=head2 errstr

When an error occurs, you can retrieve the error message either from the
C<$DB::Color::Config::errstr> variable, or using the C<errstr()> method.

=head1 CAVEATS

=head2 Unsupported Section Headers

Some edge cases in section headers are not support, and additionally may not
be detected when writing the config file.

Specifically, section headers with leading whitespace, trailing whitespace,
or newlines anywhere in the section header, will not be written correctly
to the file and may cause file corruption.

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Config-Tiny>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 ACKNOWLEGEMENTS

Thanks to Sherzod Ruzmetov E<lt>sherzodr@cpan.orgE<gt> for
L<Config::Simple>, which inspired this module by being not quite
"simple" enough for me :)

=head1 SEE ALSO

L<Config::Simple>, L<Config::General>, L<ali.as>

=head1 COPYRIGHT

Copyright 2002 - 2011 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


lib/DB/Color/Highlight.pm
package DB::Color::Highlight;

use strict;
use warnings;
use Term::ANSIColor ':constants';
use Digest::MD5 'md5_hex';
use File::Spec::Functions qw(catfile catdir);
use File::Path 'make_path';

BEGIN {
    if ( !( Term::ANSIColor->VERSION >= 3 ) ) {
        no warnings 'redefine';
        *BRIGHT_BLUE = sub { BLUE };
    }
}

use Syntax::Highlight::Engine::Kate::Perl;

=head1 NAME

DB::Color::Highlight - Provides highlighting for DB::Color

=head1 VERSION

Version 0.20

=cut

our $VERSION = '0.20';

# increase this number by one to force the cache to generate new md5 numbers
my $FORMAT_NUMBER = 1;

BEGIN {
    no warnings 'redefine';
    *Syntax::Highlight::Engine::Kate::Template::logwarning = sub { };
}

sub new {
    my ( $class, $args ) = @_;
    my $self = bless {} => $class;
    $self->_initialize($args);
    return $self;
}

sub _initialize {
    my ( $self, $args ) = @_;

    my $cache_dir = $args->{cache_dir};
    $self->{debug_fh}  = $args->{debug_fh};
    $self->{cache_dir} = $cache_dir;

    if ( defined $cache_dir and not -d $cache_dir ) {
        mkdir $cache_dir or die "Cannot mkdir ($cache_dir): $!";
    }

#      CLEAR           RESET             BOLD            DARK
#      FAINT           ITALIC            UNDERLINE       UNDERSCORE
#      BLINK           REVERSE           CONCEALED
#
#      BLACK           RED               GREEN           YELLOW
#      BLUE            MAGENTA           CYAN            WHITE
#      BRIGHT_BLACK    BRIGHT_RED        BRIGHT_GREEN    BRIGHT_YELLOW
#      BRIGHT_BLUE     BRIGHT_MAGENTA    BRIGHT_CYAN     BRIGHT_WHITE
#
#      ON_BLACK        ON_RED            ON_GREEN        ON_YELLOW
#      ON_BLUE         ON_MAGENTA        ON_CYAN         ON_WHITE
#      ON_BRIGHT_BLACK ON_BRIGHT_RED     ON_BRIGHT_GREEN ON_BRIGHT_YELLOW
#      ON_BRIGHT_BLUE  ON_BRIGHT_MAGENTA ON_BRIGHT_CYAN  ON_BRIGHT_WHITE

    my $highlighter = Syntax::Highlight::Engine::Kate::Perl->new(
        format_table => {
            'Keyword'      => [ YELLOW,      RESET ],
            'Comment'      => [ BRIGHT_BLUE, RESET ],
            'Decimal'      => [ YELLOW,      RESET ],
            'Float'        => [ YELLOW,      RESET ],
            'Function'     => [ CYAN,        RESET ],
            'Identifier'   => [ RED,         RESET ],
            'Normal'       => [ WHITE,       RESET ],
            'Operator'     => [ CYAN,        RESET ],
            'Preprocessor' => [ RED,         RESET ],
            'String'       => [ MAGENTA,     RESET ],
            'String Char'  => [ RED,         RESET ],
            'Symbol'       => [ CYAN,        RESET ],
            'DataType'     => [ CYAN,        RESET ],    # variable names
        }
    );
    $self->{highlighter} = $highlighter;
}

sub _highlighter  { $_[0]->{highlighter} }
sub _cache_dir    { $_[0]->{cache_dir} }
sub _should_cache { defined $_[0]->_cache_dir }

sub _debug {
    my ( $self, $message ) = @_;
    return unless my $debug = $self->{debug_fh};
    print $debug "$message\n";
}

sub highlight_text {
    my ( $self, $code ) = @_;

    if ( $self->_should_cache ) {
        my ( $path, $file ) = $self->_get_path_and_file($code);
        unless ( -d $path ) {
            make_path($path);
        }
        $file = catfile( $path, $file );

        if ( -e $file ) {
            $self->_debug("Cache hit on '$file'");

            # update the atime, mtime to ensure that our naive cache recognizes
            # this as a "recent" file
            utime time, time, $file or die "Cannot 'utime atime, mtime $file: $!";
            open my $fh, '<', $file or die "Cannot open '$file' for reading: $!";
            return do { local $/; <$fh> };
        }
        else {
            $self->_debug("Cache miss on '$file'");
            my $highlighted = $self->_get_highlighted_text($code);
            open my $fh, '>', $file or die "Cannot open '$file' for writing: $!";
            print $fh $highlighted;
            return $highlighted;
        }
    }
    else {
        return $self->_get_highlighted_text($code);
    }
}

sub _get_highlighted_text {
    my ( $self, $code ) = @_;

    my @code;
    my $line_num = 0;
    my $in_pod   = 0;
    my %pod_lines;
    my @pod_line_nums;
    foreach ( split /\n/ => $code ) {
        if (/^=(?!cut\b)/) {
            $in_pod = 1;
        }
        if ($in_pod) {
            $pod_lines{$line_num} = $_;
            push @pod_line_nums => $line_num;
            push @code          => '';
        }
        else {
            push @code => $_;
        }
        if (/^=cut\b/) {
            $in_pod = 0;
        }
        $line_num++;
    }
    $code = join "\n" => @code;
    my $highlighted = $self->_highlighter->highlightText($code);
    @code = split /\n/ => $highlighted;
    @code[@pod_line_nums] = @pod_lines{@pod_line_nums};
    return join "\n" => map { BLUE . $_ . RESET } @code;
}

sub _get_path_and_file {
    my ( $self, $code ) = @_;
    unless ( $self->_should_cache ) {
        $self->_debug("Caching disabled");
        return;
    }
    my $md5 = md5_hex( $self->_get_unique_factors, $code );
    my $dir = substr $md5, 0, 2, '';
    my $file = $md5;

    my $path = catdir( $self->_cache_dir, $dir );
    $self->_debug("Cache path is '$path'. Cache file is '$file'");
    return $path, $file;
}

sub _format_number {
    return $FORMAT_NUMBER;
}

sub _get_unique_factors {
    my $self = shift;
    return ( $self->_format_number, ref $self );
}

1;
__END__

=head1 SYNOPSIS

 use DB::Color::Highlight;
 my $highlighter = DB::Color::Highlight::highlighter();
 my $highlighted = $highlighter->highlightText($code);

=head1 INTERNAL USE ONLY

Don't touch this. It's subject to change at any time.

=head1 EXPORT

Nothing.

=head1 SUBROUTINES

=head2 C<highlighter>

Returns a L<Syntax::Highlight::Engine::Kate::Perl> object.

=cut


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name       => 'DB::Color',
    license           => 'perl',
    dist_author       => q{Curtis 'Ovid' Poe <ovid@cpan.org>},
    dist_version_from => 'lib/DB/Color.pm',
    build_requires    => { 'Test::More' => 0, },
    requires          => {
        'Syntax::Highlight::Engine::Kate' => 0.06,
        'Digest::MD5'                     => 0,
        'Term::ANSIColor' => 0,        # 3.0 or better to get bright colors
        'File::Path'      => '2.07',
    },
    configure_requires => { 'Module::Build' => '0.4' },
    add_to_cleanup       => ['DB-Color-*'],
    create_makefile_pl   => 'traditional',
    recursive_test_files => 1,
    script_files         => 'script',
    meta_merge =>
      { resources => { repository => 'https://github.com/Ovid/DB--Color' } },
);

$builder->create_build_script();
